core/modules/wpublish/QueriesAndChunks.cc

```diff
@@ -0,0 +1,486 @@
+        for (auto const& ele : _queryStats) {
+            auto const& q = ele.second;
+            uqs.push_back(q);
+        }
+    }
+
+    // Go through all Tasks in each query and examine the running ones.
+    // If a running Task is taking longer than its percent of total time, boot it.
+    // Booting a Task may result in the entire user query it belongs to being moved
+    // to the snail scan.
+    for (auto const& uq : uqs) {
+        // Copy all the running tasks that are on ScanSchedulers.
+        std::vector<wbase::Task::Ptr> runningTasks;
+        {
+            std::lock_guard<std::mutex> lock(uq->_qStatsMtx);
+            for (auto const& ele : uq->_taskMap) {
+                auto const& task = ele.second;
+                if (task->getState() == wbase::Task::State::RUNNING) {
+                    auto const& sched = std::dynamic_pointer_cast<wsched::ScanScheduler>(task->getTaskScheduler());
+                    if (sched != nullptr) {
+                        runningTasks.push_back(task);
+                    }
+                }
+            }
+        }
+
+        // For each running task, check if it is taking too long, or if the query is taking too long.
+        for (auto const& task : runningTasks) {
+            auto const& sched = std::dynamic_pointer_cast<wsched::ScanScheduler>(task->getTaskScheduler());
+            if (sched == nullptr) {
+                continue;
+            }
+            double schedMaxTime = sched->getMaxTimeMinutes(); // Get max time for scheduler
+            // Get the slowest scan table in task.
+            auto begin = task->getScanInfo().infoTables.begin();
+            if (begin == task->getScanInfo().infoTables.end()) {
+                continue;
+            }
+            std::string const& slowestTable = begin->db + ":" + begin->table;
+            auto iterTbl = scanTblSums.find(slowestTable);
+            if (iterTbl != scanTblSums.end()) {
+                LOGS(_log, LOG_LVL_DEBUG, "examineAll " << slowestTable
+                                       << " chunkId=" << task->getChunkId());
+                ScanTableSums &tblSums = iterTbl->second;
+                auto iterChunk = tblSums.chunkPercentages.find(task->getChunkId());
+                if (iterChunk != tblSums.chunkPercentages.end()) {
+                    // We can only make the check if there's data on past chunks/tables.
+                    double percent = iterChunk->second.percent;
+                    double maxTimeChunk = percent * schedMaxTime;
+                    auto runTimeMilli = task->getRunTime();
+                    double runTimeMinutes = (double)runTimeMilli.count()/60000.0;
+                    std::ostringstream os;
+                    os << task->getIdStr()
+                       << "maxTimeChunk(" << maxTimeChunk << ")=percent(" << percent << ")*schedMaxTime(" << schedMaxTime << ")"
+                       << " runTimeMinutes=" << runTimeMinutes;
```

Comment:
Could you avoid unconditional formatting to a stream? Something like this might work:

    bool booting = runTimeMinutes > maxTimeChunk;
    auto lvl = booting ? LOG_LVL_INFO : LOG_LVL_DEBUG;
    LOGS(_log, lvl, "examineAll " << (booting ? "booting" : "keeping") << " task " << task->getIdStr() << ...);